Cluster panic

Some time after start, all nodes of my cluster (two servers and two agents) crash with the same error:

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x160 pc=0x948a8a]

goroutine 623378 [running]:
google.golang.org/grpc.(*Server).Stop(0x0)
        /home/runner/go/pkg/mod/google.golang.org/grpc@v1.29.1/server.go:1482 +0x4a
github.com/distribworks/dkron/v3/plugin.(*ExecutorClient).Execute(0xc00091e6a0, 0xc00b231640, 0x357c300, 0xc000e27820, 0xc0011bc980, 0x25bb801, 0xc0044c6150)
        /home/runner/work/dkron/dkron/plugin/executor.go:65 +0x190
github.com/distribworks/dkron/v3/dkron.(*AgentServer).AgentRun(0xc0001295b8, 0xc0044c6060, 0x35e2620, 0xc003374e20, 0x0, 0x0)
        /home/runner/work/dkron/dkron/dkron/grpc_agent.go:83 +0x6f4
github.com/distribworks/dkron/v3/plugin/types._Agent_AgentRun_Handler(0x2561d80, 0xc0001295b8, 0x35dcda0, 0xc0002186c0, 0x4ba5ae0, 0xc000e2a200)
        /home/runner/work/dkron/dkron/plugin/types/dkron.pb.go:1862 +0x109
google.golang.org/grpc.(*Server).processStreamingRPC(0xc000a12b60, 0x35ea7e0, 0xc0019b4300, 0xc000e2a200, 0xc000b56150, 0x4b45e20, 0x0, 0x0, 0x0)
        /home/runner/go/pkg/mod/google.golang.org/grpc@v1.29.1/server.go:1329 +0xcbc
google.golang.org/grpc.(*Server).handleStream(0xc000a12b60, 0x35ea7e0, 0xc0019b4300, 0xc000e2a200, 0x0)
        /home/runner/go/pkg/mod/google.golang.org/grpc@v1.29.1/server.go:1409 +0xc64
google.golang.org/grpc.(*Server).serveStreams.func1.1(0xc001ab37e0, 0xc000a12b60, 0x35ea7e0, 0xc0019b4300, 0xc000e2a200)
        /home/runner/go/pkg/mod/google.golang.org/grpc@v1.29.1/server.go:746 +0xa1
created by google.golang.org/grpc.(*Server).serveStreams.func1
        /home/runner/go/pkg/mod/google.golang.org/grpc@v1.29.1/server.go:744 +0xa1
runtime: note: your Linux kernel may be buggy
runtime: note: see https://golang.org/wiki/LinuxKernelSignalVectorBug
runtime: note: mlock workaround for kernel bug failed with errno 12
```

### Environment

- Dkron 3.0.5
- Docker (servers: [Dockerfile](https://github.com/alexanderfefelov/docker-backpack/blob/master/workload-automation/dkron/server/Dockerfile), [run script](https://github.com/alexanderfefelov/docker-backpack/blob/master/workload-automation/dkron/server/functions.sh), agents: [Dockerfile](https://github.com/alexanderfefelov/docker-backpack/blob/master/workload-automation/dkron/agent/Dockerfile), [run script](https://github.com/alexanderfefelov/docker-backpack/blob/master/workload-automation/dkron/agent/functions.sh))
- `uname -a`:

        Linux dkron-server-1.backpack.test 5.4.0-48-generic #52-Ubuntu SMP Thu Sep 10 10:58:49 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux

We are observing very similar issue with our cluster of 3 servers. We are using 
* dkron/dkron:v3.0.4
```
uname -a:
Linux dkron-cluster-0 4.15.0-1027-aws #27-Ubuntu SMP Fri Nov 2 15:14:20 UTC 2018 x86_64 Linux
```

Observed panic seems to originate from here https://github.com/distribworks/dkron/blob/19c09828adba4789255d92069613bfb7b5ce2395/plugin/executor.go#L65

as a result our cluster stops to execute jobs, despite that this panic happens to only one of the nodes.

the error in logs looks like:
```
/go/pkg/mod/google.golang.org/grpc@v1.29.1/server.go:744 +0xa1
--
created by google.golang.org/grpc.(*Server).serveStreams.func1
/go/pkg/mod/google.golang.org/grpc@v1.29.1/server.go:746 +0xa1
google.golang.org/grpc.(*Server).serveStreams.func1.1(0xc000edaa00, 0xc0003f2680, 0x357f1a0, 0xc00109ca80, 0xc000ef1800)
/go/pkg/mod/google.golang.org/grpc@v1.29.1/server.go:1409 +0xc64
google.golang.org/grpc.(*Server).handleStream(0xc0003f2680, 0x357f1a0, 0xc00109ca80, 0xc000ef1800, 0x0)
/go/pkg/mod/google.golang.org/grpc@v1.29.1/server.go:1329 +0xcbc
google.golang.org/grpc.(*Server).processStreamingRPC(0xc0003f2680, 0x357f1a0, 0xc00109ca80, 0xc000ef1800, 0xc000cf4780, 0x4aaedc0, 0x0, 0x0, 0x0)
/dkron/plugin/types/dkron.pb.go:1862 +0x109
github.com/distribworks/dkron/v3/plugin/types._Agent_AgentRun_Handler(0x252a1c0, 0xc00000f040, 0x35717c0, 0xc0001ac600, 0x4b0ea40, 0xc000ef1800)
/dkron/dkron/grpc_agent.go:83 +0x6f4
github.com/distribworks/dkron/v3/dkron.(*AgentServer).AgentRun(0xc00000f040, 0xc000d384e0, 0x3577040, 0xc000cc8610, 0x0, 0x0)
/dkron/plugin/executor.go:65 +0x190
github.com/distribworks/dkron/v3/plugin.(*ExecutorClient).Execute(0xc000ba8c20, 0xc000f34780, 0x3511180, 0xc0012c6d20, 0xc000e2a800, 0xc000104501, 0x101)
/go/pkg/mod/google.golang.org/grpc@v1.29.1/server.go:1482 +0x4a
google.golang.org/grpc.(*Server).Stop(0x0)
goroutine 2579 [running]:
{"service":"dkron","journald":{"_PID":"814","_SYSTEMD_INVOCATION_ID":"00e226322d7d42daa35605c8209f8cc9","_CAP_EFFECTIVE":"3fffffffff","_SOURCE_REALTIME_TIMESTAMP":"1604214389481076","PRIORITY":"3","CONTAINER_NAME":"k8s_dkron_dkron-cluster-0_tfprod_78ace7ae-1b7f-11eb-9a01-029fce2970d4_19","_SYSTEMD_CGROUP":"/system.slice/docker.service","_GID":"0","_MACHINE_ID":"3c325c220068e72790ee19b05f9d5d88","_SELINUX_CONTEXT":"unconfined\n","_COMM":"dockerd","CONTAINER_ID_FULL":"9845f9bc1c140eb037c22335b0d4bad8ddc3696f1d004b6dcff5ff9ee20d4241","CONTAINER_TAG":"9845f9bc1c14","_BOOT_ID":"2cc14745820148a4ab685328775b0890","_SYSTEMD_UNIT":"docker.service","_TRANSPORT":"journal","_CMDLINE":"/usr/bin/dockerd -H fd:// --containerd=/run/containerd/containerd.sock","SYSLOG_IDENTIFIER":"9845f9bc1c14","_SYSTEMD_SLICE":"system.slice","_UID":"0","_HOSTNAME":"kubeworker-7","CONTAINER_ID":"9845f9bc1c14","_EXE":"/usr/bin/dockerd-ce"}}
[signal SIGSEGV: segmentation violation code=0x1 addr=0x160 pc=0x9488fa]
panic: runtime error: invalid memory address or nil pointer dereference
```

Is there maybe some known configuration setting that could prevent cluster to go down in such scenario? Or any timeline when #835 could be merged and released?

Agent crash on 3.0.8:

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x160 pc=0x94baaa]

goroutine 20660 [running]:
google.golang.org/grpc.(*Server).Stop(0x0)
        /home/runner/go/pkg/mod/google.golang.org/grpc@v1.29.1/server.go:1482 +0x4a
github.com/distribworks/dkron/v3/plugin.(*ExecutorClient).Execute(0xc000c32280, 0xc00060cf80, 0x3581720, 0xc00000d5a0, 0xc00013ab00, 0x25c0201, 0xc00067f8c0)
        /home/runner/work/dkron/dkron/plugin/executor.go:65 +0x190
github.com/distribworks/dkron/v3/dkron.(*AgentServer).AgentRun(0xc00012a1c0, 0xc00067f710, 0x35e7c60, 0xc000cdebc0, 0x0, 0x0)
        /home/runner/work/dkron/dkron/dkron/grpc_agent.go:83 +0x6f4
github.com/distribworks/dkron/v3/plugin/types._Agent_AgentRun_Handler(0x2566820, 0xc00012a1c0, 0x35e23e0, 0xc000b8acc0, 0x4bacc00, 0xc0000e2b00)
        /home/runner/work/dkron/dkron/plugin/types/dkron.pb.go:1862 +0x109
google.golang.org/grpc.(*Server).processStreamingRPC(0xc0008f8340, 0x35efd60, 0xc0009d8f00, 0xc0000e2b00, 0xc000b1f1d0, 0x4b4cf20, 0x0, 0x0, 0x0)
        /home/runner/go/pkg/mod/google.golang.org/grpc@v1.29.1/server.go:1329 +0xcbc
google.golang.org/grpc.(*Server).handleStream(0xc0008f8340, 0x35efd60, 0xc0009d8f00, 0xc0000e2b00, 0x0)
        /home/runner/go/pkg/mod/google.golang.org/grpc@v1.29.1/server.go:1409 +0xc64
google.golang.org/grpc.(*Server).serveStreams.func1.1(0xc000055820, 0xc0008f8340, 0x35efd60, 0xc0009d8f00, 0xc0000e2b00)
        /home/runner/go/pkg/mod/google.golang.org/grpc@v1.29.1/server.go:746 +0xa1
created by google.golang.org/grpc.(*Server).serveStreams.func1
        /home/runner/go/pkg/mod/google.golang.org/grpc@v1.29.1/server.go:744 +0xa1
```

I have the exact same issue as well. 

- Dkron cluster of `3 x n2-standard-4 (4 vCPUs, 16 GB memory)` (Google Cloud)
- Jobs
  - Schedule: `@every 1s`
  - Executor: HTTP
  - HTTP Callback: `POST http://job-worker-host:8081/job?job_id={jobID}` (endpoint simply emits a statsd counter)
  - Number of Jobs: 100

The expectation was to see 100 rps sustained throughput on the `/job` endpoint. But one or all nodes crash with following panic after sometime (all jobs completed about 20 executions)

```
INFO[2021-02-22T21:30:29+07:00] grpc_agent: Starting job                      job=test_j98 node=p-dkron-node-a-01
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x168 pc=0x9535ca]

goroutine 423018 [running]:
google.golang.org/grpc.(*Server).Stop(0x0)
        /home/runner/go/pkg/mod/google.golang.org/grpc@v1.35.0/server.go:1626 +0x4a
github.com/distribworks/dkron/v3/plugin.(*ExecutorClient).Execute(0xc0003894a0, 0xc002105860, 0x3b177a0, 0xc000eff240, 0xc0005bd950, 0x1, 0x0)
        /home/runner/work/dkron/dkron/plugin/executor.go:65 +0x191
github.com/distribworks/dkron/v3/dkron.(*AgentServer).AgentRun(0xc00101bb20, 0xc0019675c0, 0x3b88be0, 0xc003f5c650, 0x0, 0x0)
        /home/runner/work/dkron/dkron/dkron/grpc_agent.go:85 +0x6fb
github.com/distribworks/dkron/v3/plugin/types._Agent_AgentRun_Handler(0x2969a20, 0xc00101bb20, 0x3b82be0, 0xc0023c9740, 0x535e610, 0xc0018d8c00)
        /home/runner/work/dkron/dkron/plugin/types/dkron_grpc.pb.go:540 +0x109
google.golang.org/grpc.(*Server).processStreamingRPC(0xc000130c40, 0x3b91700, 0xc002f4f080, 0xc0018d8c00, 0xc0009484b0, 0x52f76e0, 0x0, 0x0, 0x0)
        /home/runner/go/pkg/mod/google.golang.org/grpc@v1.35.0/server.go:1464 +0xcbd
google.golang.org/grpc.(*Server).handleStream(0xc000130c40, 0x3b91700, 0xc002f4f080, 0xc0018d8c00, 0x0)
        /home/runner/go/pkg/mod/google.golang.org/grpc@v1.35.0/server.go:1544 +0xc96
google.golang.org/grpc.(*Server).serveStreams.func1.2(0xc002514e50, 0xc000130c40, 0x3b91700, 0xc002f4f080, 0xc0018d8c00)
        /home/runner/go/pkg/mod/google.golang.org/grpc@v1.35.0/server.go:878 +0xa1
created by google.golang.org/grpc.(*Server).serveStreams.func1
        /home/runner/go/pkg/mod/google.golang.org/grpc@v1.35.0/server.go:876 +0x204
```

--- a/plugin/executor.go
+++ b/plugin/executor.go
@@ -37,11 +37,16 @@ func (p *ExecutorPlugin) GRPCClient(ctx context.Context, broker *plugin.GRPCBrok
 	return &ExecutorClient{client: types.NewExecutorClient(c), broker: broker}, nil
 }
 
+type Broker interface {
+	NextId() uint32
+	AcceptAndServe(id uint32, s func([]grpc.ServerOption) *grpc.Server)
+}
+
 // Here is the gRPC client that GRPCClient talks to.
 type ExecutorClient struct {
 	// This is the real implementation
 	client types.ExecutorClient
-	broker *plugin.GRPCBroker
+	broker Broker
 }
 
 func (m *ExecutorClient) Execute(args *types.ExecuteRequest, cb StatusHelper) (*types.ExecuteResponse, error) {

--- a/plugin/executor.go
+++ b/plugin/executor.go
@@ -48,16 +48,26 @@ func (m *ExecutorClient) Execute(args *types.ExecuteRequest, cb StatusHelper) (*
 	// This is where the magic conversion to Proto happens
 	statusHelperServer := &GRPCStatusHelperServer{Impl: cb}
 
+	initChan := make(chan bool, 1)
 	var s *grpc.Server
 	serverFunc := func(opts []grpc.ServerOption) *grpc.Server {
 		s = grpc.NewServer(opts...)
 		types.RegisterStatusHelperServer(s, statusHelperServer)
+		initChan <- true
 
 		return s
 	}
 
 	brokerID := m.broker.NextId()
-	go m.broker.AcceptAndServe(brokerID, serverFunc)
+	go func() {
+		m.broker.AcceptAndServe(brokerID, serverFunc)
+		// AcceptAndServe might terminate without calling serverFunc
+		// To prevent eternal blocking, send 'init done' signal
+		initChan <- true
+	}()
+
+	// Wait for s to be initialized in the goroutine
+	<-initChan
 
 	args.StatusServer = brokerID
 	r, err := m.client.Execute(context.Background(), args)

--- a/plugin/executor_test.go
+++ b/plugin/executor_test.go
@@ -1,0 +1,47 @@
+package plugin
+
+import (
+	"context"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+
+	"github.com/distribworks/dkron/v3/plugin/types"
+	dktypes "github.com/distribworks/dkron/v3/plugin/types"
+	grpc "google.golang.org/grpc"
+)
+
+type MockedExecutor struct{}
+
+func (m *MockedExecutor) Execute(ctx context.Context, in *types.ExecuteRequest, opts ...grpc.CallOption) (*types.ExecuteResponse, error) {
+	resp := &dktypes.ExecuteResponse{}
+	return resp, nil
+}
+
+type MockedStatusHelper struct{}
+
+func (m MockedStatusHelper) Update([]byte, bool) (int64, error) {
+	return 0, nil
+}
+
+type MockedBroker struct{}
+
+func (m *MockedBroker) AcceptAndServe(id uint32, s func([]grpc.ServerOption) *grpc.Server) {
+}
+
+func (m *MockedBroker) NextId() uint32 {
+	return 0
+}
+
+func TestExecuteDoesNotPanicIfGRPCIsNotInitializedOnTime(t *testing.T) {
+	var brokerMock MockedBroker
+	var execMock MockedExecutor
+	execClient := ExecutorClient{
+		client: &execMock,
+		broker: &brokerMock,
+	}
+
+	var requestStub dktypes.ExecuteRequest
+	var statusHelperMock MockedStatusHelper
+	assert.NotPanics(t, func() { execClient.Execute(&requestStub, statusHelperMock) })
+}

